src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs

```diff
@@ -3154,22 +3148,106 @@ public static void Sort<T>(this Span<T> span, Comparison<T> comparison)
                         Unsafe.As<T, int>(ref oldValue),
                         Unsafe.As<T, int>(ref newValue),
                         length);
+                    return;
                 }
-                else
+                else if (sizeof(T) == sizeof(long))
                 {
-                    Debug.Assert(sizeof(T) == sizeof(long));
-
                     ref long src = ref Unsafe.As<T, long>(ref MemoryMarshal.GetReference(span));
                     SpanHelpers.ReplaceValueType(
                         ref src,
                         ref src,
                         Unsafe.As<T, long>(ref oldValue),
                         Unsafe.As<T, long>(ref newValue),
                         length);
+                    return;
+                }
+            }
+
+            ref T src2 = ref MemoryMarshal.GetReference(span);
+            SpanHelpers.Replace(ref src2, ref src2, oldValue, newValue, length);
+        }
+
+        /// <summary>
+        /// Copies <paramref name="source"/> to <paramref name="destination"/>, replacing all occurrences of <paramref name="oldValue"/> with <paramref name="newValue"/>.
+        /// </summary>
+        /// <typeparam name="T">The type of the elements in the spans.</typeparam>
+        /// <param name="source">The span to copy.</param>
+        /// <param name="destination">The span into which the copied and replaced values should be written.</param>
+        /// <param name="oldValue">The value to be replaced with <paramref name="newValue"/>.</param>
+        /// <param name="newValue">The value to replace all occurrences of <paramref name="oldValue"/>.</param>
+        /// <exception cref="ArgumentException">The <paramref name="destination"/> span was shorter than the <paramref name="source"/> span.</exception>
+        /// <exception cref="ArgumentException">The <paramref name="source"/> and <paramref name="destination"/> were overlapping but not referring to the same starting location.</exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe void Replace<T>(this ReadOnlySpan<T> source, Span<T> destination, T oldValue, T newValue) where T : IEquatable<T>?
+        {
+            nuint length = (uint)source.Length;
+            if (length == 0)
+            {
+                return;
+            }
+
+            if (length > (uint)destination.Length)
+            {
+                ThrowHelper.ThrowArgumentException_DestinationTooShort();
+            }
+
+            ref T src = ref MemoryMarshal.GetReference(source);
+            ref T dst = ref MemoryMarshal.GetReference(destination);
+
+            nint byteOffset = Unsafe.ByteOffset(ref src, ref dst);
+            if (byteOffset != 0 &&
+                ((nuint)byteOffset < (nuint)((nint)source.Length * sizeof(T)) ||
+                 (nuint)byteOffset > (nuint)(-((nint)destination.Length * sizeof(T)))))
```

Comment:
```suggestion
            if (source.Overlaps(destination))
```
instead of manually inlining of that method? At the moment the [codegen](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4a9Jq049+ggJI8ZEMfKhKV6zY1kALbFDEAZbMHbc+AjdRoBiLg5JSTdpBgATXlxQmBoaAG0AKV4MAHEYLhllAAoMAE8xGAgAM2zgPIwYAEoqgF1NAGYGbmjimCZSBiEAHgAVAD4GAG8aBjGmJuAICEkGAFlsthhscIB5Lkk82TFsLj7B3AgOKBU0Bm3d/YjVPi5sPgguKoZR8beAXgOjk8FVhRkQmJcNlwjdeHcHk8fG9XmNiJNprMAHKLZZrDZbHZ7AYMQ7HU7nLFXUG4W73XiPKqw4bUt6wYoMXq4k4Md4MenzGD4aB5OaOXAOSQsdIYJZtWBcFTZPE/KrQt7jDlM8Kk1nsmAMuZcnl8qAC7BCkVimQZKUksmQuW08bWsZccEYBjlSqrYrFXAwR1sgCqXFaggAQhUYK73Z7shzcCczhyVRgrdQFTbE0mmAB2J3B0Mex0AQjZjAAZIXbamxtlsoEHVVnSG3TmGN0GJWONWK/aeFUZSoWM4MgBzDB2BgAKlxvAAXkVSr0agxCIRS2Xm1XO7Xs56GIMW224O3q+bweTHr2B0PR+OpyVsrOagm3gBfGgPoA) for `Overlaps` shows a call, but this can be fixed (at least I hope so).

Reply:
That wouldn't be the same: it's ok for the src and dst to overlap if they start at the same location. We could use both !AreSame and Overlaps, but then we're doing extra work, like the empty tests for both src and dst in Overlaps.